### PR TITLE
Make event notifiers able to pass info to listeners

### PR DESCRIPTION
**Pull request by tbursztyka**
Let's normalize which event send what information.
ADD/DEL addr would be an in_addr/in6_addr (depending on IP version)
Route would need a bit more? Addr + prefix ?

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

### CODE REVIEW

File: include/net/net_mgmt.h [hunk 1]
```diff
@@ -189,8 +193,24 @@ void net_mgmt_del_event_callback(struct net_mgmt_event_callback *cb);
...
+ *        with the event. NULL otherwise. Note the data pointed there is
+ *        normalized by the related event.
+ * @param length size of the data pointed by info pointer.
+ *
+ * Note: info and length are disabled if CONFIG_NET_MGMT_EVENT_INFO
+ *       is not defined.
  */
-void net_mgmt_event_notify(u32_t mgmt_event, struct net_if *iface);
+#ifdef CONFIG_NET_MGMT_EVENT_INFO
+void net_mgmt_event_notify(u32_t mgmt_event, struct net_if *iface,
```
Comment (jukkar) [inline include/net/net_mgmt.h]: I think it is better to call this net_mgmt_event_notify_info() or similar, so that we can avoid having ifdefs in the code that  uses this function. We could even make the net_mgmt_event_notify() to call the new function with NULL info.

File: include/net/net_mgmt.h [hunk 2]
```diff
@@ -212,6 +234,7 @@ void net_mgmt_event_notify(u32_t mgmt_event, struct net_if *iface);
 int net_mgmt_event_wait(u32_t mgmt_event_mask,
 			u32_t *raised_event,
 			struct net_if **iface,
+			const void **info,
```
Comment (jukkar) [inline include/net/net_mgmt.h]: Don't we need the length too here?

File: include/net/net_mgmt.h [hunk 3]
```diff
@@ -233,6 +258,7 @@ int net_mgmt_event_wait(u32_t mgmt_event_mask,
 int net_mgmt_event_wait_on_iface(struct net_if *iface,
 				 u32_t mgmt_event_mask,
 				 u32_t *raised_event,
+				 const void **info,
```
Comment (jukkar) [inline include/net/net_mgmt.h]: Same here.

File: subsys/net/ip/Kconfig.mgmt [hunk 4]
```diff
@@ -23,18 +23,18 @@ config NET_MGMT_EVENT
 	(or else) needs to be notified on a specific network event
 	(ip address change for instance) to trigger some related work.
 
+if NET_MGMT_EVENT
```
Comment (jukkar) [inline subsys/net/ip/Kconfig.mgmt]: This looks like unrelated change, perhaps create another commit for this.

File: subsys/net/ip/net_if.c [hunk 5]
```diff
@@ -443,7 +443,8 @@ static void dad_timeout(struct k_work *work)
 	 */
 	tmp = net_if_ipv6_addr_lookup(&ifaddr->address.in6_addr, &iface);
 	if (tmp == ifaddr) {
-		net_mgmt_event_notify(NET_EVENT_IPV6_DAD_SUCCEED, iface);
+		net_mgmt_event_notify(NET_EVENT_IPV6_DAD_SUCCEED,
```
Comment (jukkar) [inline subsys/net/ip/net_if.c]: Ẁe could skip all the changes to existing users of the net_mgmt_event_notify() if we have a separate function which has the info and length parameter. Then we can go through the listeners and convert them case by case separately.

File: subsys/net/ip/net_mgmt.c [hunk 6]
```diff
@@ -20,6 +20,14 @@
 struct mgmt_event_entry {
 	u32_t event;
 	struct net_if *iface;
+
+#ifdef CONFIG_NET_MGMT_EVENT_INFO
+	union {
+		u8_t info[CONFIG_NET_MGMT_EVENT_INFO_SIZE];
```
Comment (jukkar) [inline subsys/net/ip/net_mgmt.c]: This is very unfortunate, as it bloats the event entry. I wonder if we could allocate the space outside of mgmt_event_entry and only store pointer here.

File: include/net/net_mgmt.h (hunk 1 above)
Comment (tbursztyka) [reply include/net/net_mgmt.h]: hum, naming will be an issue here. It's an event we are notifying, not something else.

What about:
- net_mgmt_event_notify()  with info parameter
- net_mgmt_event_notify_only()  or net_mgmt_event_notify_no_info() ? for the non-info based notification.

File: include/net/net_mgmt.h (hunk 2 above)
Comment (tbursztyka) [reply include/net/net_mgmt.h]: No, we will document the net events so that event E meant info of this type T (and thus the implied length). It's not going to happen that 1 event can send a different type of info

File: subsys/net/ip/Kconfig.mgmt (hunk 4 above)
Comment (tbursztyka) [reply subsys/net/ip/Kconfig.mgmt]: ok

File: subsys/net/ip/net_if.c (hunk 5 above)
Comment (tbursztyka) [reply subsys/net/ip/net_if.c]: As noted previously, net_mgmt_event_notify() will be the with-info function. So changes will be required anyway for non-info based event notification (but yes, an inline one calling net_mgmt_event_notify() with NULL, 0 for info.)

File: subsys/net/ip/net_mgmt.c (hunk 6 above)
Comment (tbursztyka) [reply subsys/net/ip/net_mgmt.c]: I believe the cost of code to manage external memory placeholder vs the numbers of entries vs the number of event that do not provide info makes it ok as it is now.

As far as I can tell, the info placeholder won't be that big. Currently I set it to 8bytes, but I think it will be more like 10-12 bytes. (ipv6 address + prefix for routes for instance?) it's really not much. Multiply that with number of event entries (2 by default btw) it's quite a limited amount of memory in the end, and from code perspective it was a minor amount of lines added as you can see..

With external placeholder you would still need to be prepared that all the event queued entries provide info, i.e.: you would need as many placeholders as event entries. Now, add to that it could be those would be of the maximum size info (let's say 4 routes added one by one) so you need to be prepared to provide max size for all. Now add also the pointer to store the placeholder on the entry (4bytes more). In the end, with current use case, it would be eating more memory.

File: include/net/net_mgmt.h (hunk 1 above)
Comment (tbursztyka) [reply include/net/net_mgmt.h]: actually:
- net_mgmt_event_notify()  no info parameter
- net_mgmt_event_notify_with_info() 

sound better, less changes.

File: subsys/net/ip/net_mgmt.c (hunk 6 above)
Comment (jukkar) [reply subsys/net/ip/net_mgmt.c]: Lets see how it goes, the amount of data will be much higher as IPv6 address is 16 bytes, and routes are typically 16+1.

**Comment by jukkar** [comment]
BTW, could you also create unit test that tests this.

File: subsys/net/ip/net_mgmt.c (hunk 6 above)
Comment (tbursztyka) [reply subsys/net/ip/net_mgmt.c]: indeed 16bytes, but that's still quite limited imo.

**Comment by tbursztyka** [comment]
@jukkar yep that's the plan if the RFC is accepted as it is (i.e. if it's ok to propose it as a normal PR, not as an RFC, just a title change actually)

File: subsys/net/ip/net_mgmt.c (hunk 6 above)
Comment (tbursztyka) [reply subsys/net/ip/net_mgmt.c]: As said in my previous reply: with internal or external memory allocation we have to make sure the worst case works. Worst case being: all queued events use the biggest size of info.

So in the end the memory allocation is the same (we do not share our reserved pools with anything else), being in fact a bit worse for external method since each entry needs an attributes pointing to the allocated memory.

I don't think we'll ever get huge amount of data to pass, nor huge size of entry queue in net_mgmt, so in the end it's just fine. I mean, we are talking of entry made of 4+4 bytes currently + 4 + probably 20 (16+1, with alignment it will be 20): 32 bytes. It's nothing compared to the current net buffer which size is a minimal of 128bytes. And we are not supposed to have a large amount of entries also.

Plus, we can play on the mgmt thread priority to un-queue events faster if required, to lower the need of larger queue. In any case, about this event queue size: net mgmt is not meant to be bullet proof guarantied on message passing. If the system is over-loaded, on a full event queue: the oldest will get scraped by a new one. A bit like netlink in a way, if the system is overloaded, the kernel might scrap old message in the socket buffer. There isn't much we can do actually.

Anyway, if really necessary, once Johan has done the split of net_buf metadata/data, we may use the memory pool of net_buf's data as an external allocator. That could be a solution if we really need to pass much bigger set of data as "event information".

File: tests/net/mgmt/src/mgmt.c [hunk 7]
```diff
@@ -34,6 +35,8 @@ static struct net_mgmt_event_callback rx_cb;
 static struct in6_addr addr6 = { { { 0xfe, 0x80, 0, 0, 0, 0, 0, 0,
 				     0, 0, 0, 0, 0, 0, 0, 0x1 } } };
 
+static char info_data[CONFIG_NET_MGMT_EVENT_INFO_SIZE] = "mgmt event info";
```
Comment (jukkar) [inline tests/net/mgmt/src/mgmt.c]: I just feel this is very error prone with global value. How could user figure out the size requirement for the info? I mean if we want to transfer ipv6 address(s) via info buffer, and some user specific data which smaller information size, then user would need to know that the info size would need to be the largest of these two. I am pretty sure user is not able to figure this out.

I know you abandoned the idea of having the info buffer / listener, but would it be possible to do it like that?

Comment (tbursztyka) [reply tests/net/mgmt/src/mgmt.c]: The event number will dictate the data:
- if up/down: none
- ipv4 add/del: struct in_addr
- ipv6 add/del: struct in6_addr
- etc...

What user specific data?

Comment (jukkar) [reply tests/net/mgmt/src/mgmt.c]: Because net_mgmt API is open, you can have your own events created and that event might have info which has some size requirements.

Comment (tbursztyka) [reply tests/net/mgmt/src/mgmt.c]: Well, up to the Kconfig option to fit the required size then. New event means new structure with a difference size, fine as long as the Kconfig option is properly set. (I could put a NET_ERR or assert actually in net_mgmt core that the notification info cannot fit into the data buffer).

net_mgmt is not netlink, it should not be used as an ipc or aggregated nested data or else. We just don't have the luxury to do that, and it's anyway not the idea behind net_mgmt. It's really meant for net core, bare minimum notification and that's it. I guess we should enforce that in the documentation.

The info buffer per listener is not possible with net_mgmt design: when the event is "notified" it's queued directly and has no knowledge of any listener. It's only when the notifier thread (internal to net_mgmt) runs that it goes through queued event and relate them to listeners relevantly. I made it like that for performance and logical purposes. (a listener might be added, right before the notifier thread gets to run for instance, etc..).
Thus why the info data is passed once from the notifier, and passed as const to the listener.

**Comment by jukkar** [comment]
> Well, up to the Kconfig option to fit the required size then. New event means new structure with a difference size, fine as long as the Kconfig option is properly set. (I could put a NET_ERR or assert actually in net_mgmt core that the notification info cannot fit into the data buffer).

We need to add more checks for the info size then, it is more likely than not that the info size will go wrong.
I just wonder if we could somehow "automate" the size calculation of the needed info buffer. I will need support for info buffer to pass information about neighbor + route additions/deletions and there the info size requirement is more than just one IPv6 address.

**Comment by jukkar** [comment]
> net_mgmt is not netlink, it should not be used as an ipc or aggregated nested data or else. We just don't have the luxury to do that, and it's anyway not the idea behind net_mgmt. It's really meant for net core, bare minimum notification and that's it. I guess we should enforce that in the documentation.

True, it is not netlink but it is still quite generic and someone will overuse it I am sure. But as you say we should document this better in the documentation.

**Comment by jukkar** [comment]
> The info buffer per listener is not possible with net_mgmt design: when the event is "notified" it's queued directly and has no knowledge of any listener. It's only when the notifier thread (internal to net_mgmt) runs that it goes through queued event and relate them to listeners relevantly. I made it like that for performance and logical purposes. (a listener might be added, right before the notifier thread gets to run for instance, etc..).
> Thus why the info data is passed once from the notifier, and passed as const to the listener.

Yes, I noticed this. It would have solved all the info size calculation issues if we could have done one info buf per listener. One option could be to place the event multiple times into internal queue, copying the info buffer per listener (so one event / listener). As there typically are not many listeners in the system, it might not be that big issue resource usage wise.

**Comment by tbursztyka** [comment]
Hum, queuing/copying per listener would be heavy in a way. Performance wise at least. Indeed, it would require 2+ listener on same events to start using too much memory, it's unlikely to happen in a way, but still possible.

The best idea, I think, would be to allocate event info on the same future mem pool (the one(s) introduced by ZEP-1835), so net mgmt event would pick necessary memory from net_buf's data mem pool. Such pool will be meant for numerous frames of variable size, and event info won't be super big I think, it's really meant for core event purpose, not dedicated info passing like a wifi scan (this in turn would work per-request, and the one raising request is the one providing memory to put the result in. A bit like what I did for 15.4 scanning)

Anyway, I'll add a size check when notifying.

**Review by jukkar — APPROVED** [review]
